Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,354 @@
+* If the `host_context_handle` represents any other context (so not the first one), these functions expose only properties from runtime configuration. These functions won't allow modification of the properties.
+
+It is possible to access runtime properties of the first initialized context in the process at any time (for reading only), by specifying `nullptr` as the `host_context_handle`.
+
+``` C
+int hostfxr_get_runtime_property(
+    const hostfxr_handle host_context_handle,
+    const char_t * name,
+    char_t * buffer,
+    size_t buffer_size,
+    size_t * buffer_used);
+```
+
+Returns the value of a property specified by its name.
+* `host_context_handle` - the initialized host context. If set to `nullptr` the function will operate on runtime properties of the first host context in the process.
+* `name` - the name of the runtime property to get. Must not be `nullptr`.
+* `buffer` - buffer to receive the value of the property
+* `buffer_size` - the size of the `buffer` in `char_t` units.
+* `buffer_used` - when successful contains number of `char_t` units used in the buffer (including the null terminator). If the specified `buffer_size` was too small, the function returns `HostApiBufferTooSmall` and sets the `buffer_used` to the minimum required size.
+
+Trying to get a property which doesn't exist is an error and will return an appropriate error code.
+
+We're proposing a fix in `hostpolicy` which will make sure that there are no duplicates possible after initialization (see dotnet/core-setup#5529). With that `hostfxr_get_runtime_property` will work always (as there can only be one value).
+
+
+``` C
+int hostfxr_set_runtime_property(
+    const hostfxr_handle host_context_handle,
+    const char_t * name,
+    const char_t * value);
+```
+
+Sets the value of a property.
+* `host_context_handle` - the initialized host context.
+* `name` - the name of the runtime property to set. Must not be `nullptr`.
+* `value` - the value of the property to set. If the property already has a value in the host context, this function will overwrite it. When set to `nullptr` and if the property already has a value then the property is "unset" - removed from the runtime property collection.
+
+Setting properties is only supported on the first host context in the process. This is really a limitation of the runtime for which the runtime properties are immutable. Once the first host context is initialized and starts a runtime there's no way to change these properties. For now we will not consider the scenario where the host context is initialized but the runtime hasn't started yet, mainly for simplicity of implementation and lack of requirements.
+
+
+``` C
+int hostfxr_get_runtime_property_count(
```

Comment:
Perhaps instead of having these two functions create a single API that accepts two buffers and the return of the internal strings?

``` C
int hostfxr_get_runtime_properties(
  const hostfxr_handle host_context_handle.
  size_t *count,
  const char_t **keys,
  const char_t **values,
);
```

Usage:

``` C
size_t property_count;

// Determine property count
rc = hostfxr_get_runtime_properties(hnd, &property_count, nullptr, nullptr);
assert(rc == small_buffer);

const char_t **key_buffer = (const char_t **)malloc(sizeof(const char_t*) * property_count);
const char_t **value_buffer = (const char_t **)malloc(sizeof(const char_t*) * property_count);

// Get all properties
rc = hostfxr_get_runtime_properties(hnd, &property_count, key_buffer, value_buffer);
assert(rc == success);

// Display properties
for (size_t i = 0; i < property_count; ++i)
{
  printf("%s=%s\n", key_buffer[i], value_buffer[i]);
}

free(key_buffer);
free(value_buffer);
```

Reply:
That's a good idea - changed to this form.